### ISSUE TITLE
Reset arm to home position

### ISSUE THREAD

**Issue by PeterGhimself** [space-concordia-robotics/robotics-prototype#28]
Use cases:
- 4.1.10 Set motors home position
The arm will move it’s motors to a constant absolute origin position which will will re-calibrate all the system variables accordingly. Due to the possible dangers of setting the motors to home position (if path is obstructed), a modal will ask the user to confirm before executing this command.

- [x] Motor stops and moves back into safe angle range when arm link hits a limit switch
- [ ] Homing sequence: Hit one or two limit switches, move to 0 degrees, optionally move next motor
- - [x] Home a single motor
- - [ ] Home all motors, one after another
- - - [x] Home all motors in open loop
- - - [ ] Home all motors in closed loop (to be fully tested, works for a single motor)
- - [x] Home motors inwards
- - [ ] Home motors both inwards and outwards (to be tested, noncritical)
- - - [ ] Use information from inner and outer homing limits to obtain better angle estimation (noncritical)
- - [ ] Choose homing direction with a command (noncritical)
- [ ] Command to move to neutral position

**Comment by vashmata** [comment]
It looks like I did a lot of work on this under the tag for manual control (#24) because I forgot we actually had a separate issue for this.... So this comment is just to indicate that homing is being worked on (and has basic functionality) in branch 24

**Comment by PeterGhimself** [comment]
@vashmata are you working on this in some branch somewhere? If so which one. Also in your checklist your last point "Command to move to neutral position" what do you mean by _neutral_ ? Is it the same as home position?

**Comment by vashmata** [comment]
Currently this is not attached to any branch. I'm defining neutral position as 0 0 0 0 0 0 and home position as the resting configuration of the arm, some configuration that looks "natural" as a starting point for competition tasks.

**Comment by PeterGhimself** [comment]
what are the advantages of having these two distinct positions to go to? Without proper encoder feedback I see how having a command to go to neutral helps us "reset" our open loop style estimated angle values, but is it as relevant with proper feedback from the motors?

**Comment by vashmata** [comment]
There isn't really an advantage to bringing the arm to "neutral" once the angles are all correct. That said this raises the question of how we should home the arm properly:
- if the whole process is automated, how do we guarantee that the arm won't accidentally collide with the ground or the rover? We'd need vision or sensors or something for self-collision and environment-collision avoidance
- otherwise my 3 ideas are:
- - we home the joints one at a time and adjust the arm's configuration as we go to ensure that it doesn't collide with anything
- - we bring the whole arm into the neutral position and then set all the joint software angles to zero
- - similar to the previous point but we go joint by joint (i.e. set one joint to 0 then reset that joint's software angle, then move to the next joint)

So basically neutral doesn't really make much sense to go to once the arm angles are known. The home position may or may not be useful. Maybe I need to rewrite this stuff a little bit.

**Comment by PeterGhimself** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [A2.2 Homing and Inverse Kinematics](https://trello.com/c/gfcSHcZL/143-a22-homing-and-inverse-kinematics)
